crates/tools/sim_chat/src/lib3h_simchat/mod.rs

```diff
@@ -0,0 +1,443 @@
+    Ok(Address::from(signature_str))
+}
+
+#[cfg(not(test))]
+pub fn current_timestamp() -> u64 {
+    std::time::SystemTime::now()
+        .duration_since(std::time::UNIX_EPOCH)
+        .unwrap()
+        .as_secs()
+}
+
+#[cfg(test)] // make sure time doesn't mess up the tests
+pub fn current_timestamp() -> u64 {
+    0
+}
+
+pub fn current_timeanchor() -> Address {
+    Address::from("0")
+}
+
+#[cfg(test)]
+mod tests {
+    pub mod mock_engine;
+
+    use super::*;
+    use crate::simchat::SimChatMessage;
+
+    fn new_sim_chat_mock_engine(callback: HandleEvent) -> Lib3hSimChat {
+        Lib3hSimChat::new(mock_engine::MockEngine::new, callback, vec![])
+    }
+
+    /*----------  example messages  ----------*/
+
+    fn chat_event() -> ChatEvent {
+        ChatEvent::SendDirectMessage {
+            to_agent: "addr".to_string(),
+            payload: "yo".to_string(),
+        }
+    }
+
+    fn receive_sys_message(payload: String) -> ChatEvent {
+        ChatEvent::ReceiveDirectMessage(SimChatMessage {
+            from_agent: String::from("sys"),
+            payload,
+            timestamp: current_timestamp(),
+        })
+    }
+
+    fn join_event() -> ChatEvent {
+        ChatEvent::Join {
+            agent_id: "test_agent".to_string(),
+            channel_id: "test_channel".to_string(),
+        }
+    }
+
+    fn join_success_event() -> ChatEvent {
+        let space_address = channel_address_from_string(&"test_channel".to_string())
+            .expect("failed to hash string");
+
+        ChatEvent::JoinSuccess {
+            channel_id: "test_channel".to_string(),
+            space_data: SpaceData {
+                agent_id: Address::from("test_agent"),
+                request_id: "".to_string(),
+                space_address,
+            },
+        }
+    }
+
+    fn part_event() -> ChatEvent {
+        ChatEvent::Part {
+            channel_id: "test_channel".to_string(),
+        }
+    }
+
+    fn part_success_event() -> ChatEvent {
+        ChatEvent::PartSuccess {
+            channel_id: "test_channel".to_string(),
+        }
+    }
+
+    #[test]
+    fn it_should_echo() {
+        let (s, r) = crossbeam_channel::unbounded();
+        let mut chat = new_sim_chat_mock_engine(Box::new(move |event| {
+            s.send(event.to_owned()).expect("send fail");
+        }));
+        chat.send(chat_event());
+        assert_eq!(chat_event(), r.recv().expect("receive fail"));
+    }
+
+    /*----------  join/part ----------*/
+
+    #[test]
+    fn calling_join_space_triggers_success_message() {
+        let (s, r) = crossbeam_channel::unbounded();
+        let mut chat = new_sim_chat_mock_engine(Box::new(move |event| {
+            s.send(event.to_owned()).expect("send fail");
+        }));
+
+        chat.send(join_event());
+
+        let chat_messages = r.iter().take(2).collect::<Vec<_>>();
+        assert_eq!(chat_messages, vec![join_event(), join_success_event(),],);
```

Comment:
These tests should not depend on the ordering of the messages as this can be changed at any time